extension/observer/cfgardenobserver/extension.go

```diff
@@ -4,33 +4,337 @@
+		infos[c.Handle()] = info
+	}
 
+	go g.updateContainerCache(infos)
 	return endpoints
 }
+
+// containerEndpoints generates a list of observer.Endpoint for a container,
+// this is because a container might have more than one exposed ports
+func (g *cfGardenObserver) containerEndpoints(handle string, info garden.ContainerInfo) []observer.Endpoint {
+	portsProp, ok := info.Properties["network.ports"]
+	if !ok {
+		g.logger.Error("could not discover container ports")
+		return nil
+	}
+	ports := strings.Split(portsProp, ",")
+
+	var app *resource.App
+	var err error
+	if g.config.IncludeAppLabels {
+		app, err = g.App(info)
+		if err != nil {
+			g.logger.Error("error fetching Application", zap.Error(err))
+			return nil
+		}
+	}
+
+	endpoints := []observer.Endpoint{}
+	for _, port := range ports {
+		port, err := strconv.ParseUint(port, 10, 16)
+		if err != nil {
+			g.logger.Error("container port is not valid", zap.Error(err))
+			continue
+		}
+
+		details := &observer.Container{
+			Name:        handle,
+			ContainerID: handle,
+			Host:        info.ContainerIP,
+			Port:        uint16(port),
+			Transport:   observer.ProtocolTCP,
+			Labels:      g.containerLabels(info, app),
+		}
+
+		endpoint := observer.Endpoint{
+			ID:      observer.EndpointID(fmt.Sprintf("%s:%d", details.ContainerID, details.Port)),
+			Target:  fmt.Sprintf("%s:%d", details.Host, details.Port),
+			Details: details,
+		}
+		endpoints = append(endpoints, endpoint)
+	}
+	return endpoints
+}
+
+func (g *cfGardenObserver) containerLabels(info garden.ContainerInfo, app *resource.App) map[string]string {
+	labels := make(map[string]string)
+	tags, err := parseTags(info)
+	if err != nil {
+		g.logger.Warn("not able to parse container tags into labels", zap.Error(err))
+		return nil
+	}
+	for k, v := range tags {
+		labels[k] = v
+	}
+
+	if app != nil {
+		for k, v := range app.Metadata.Labels {
+			labels[k] = *v
+		}
+	}
+
+	return labels
+}
+
+// The info.Properties contains a key called "log_config", which
+// has contents that look like the following JSON endoded string:
+//
+//	{
+//	  "guid": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	  "index": 0,
+//	  "source_name": "CELL",
+//	  "tags": {
+//	    "app_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	    "app_name": "example-app",
+//	    "instance_id": "0",
+//	    "organization_id": "11111111-2222-3333-4444-555555555555",
+//	    "organization_name": "example-org",
+//	    "process_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	    "process_instance_id": "abcdef12-3456-7890-abcd-ef1234567890",
+//	    "process_type": "web",
+//	    "source_id": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+//	    "space_id": "99999999-8888-7777-6666-555555555555",
+//	    "space_name": "example-space"
+//	  }
+//	}
+//
+// We parse only the tags into a map, to be used as labels
+func parseTags(info garden.ContainerInfo) (map[string]string, error) {
+	logConfig, ok := info.Properties["log_config"]
```

Comment:
Would it make sense to make `log_config` a central variable as well?

Follow up question for this PR as a whole: Does the cf golang package define any of these properties as constants somewhere that we could use, instead of defining them all ourselves?

Reply:
I have made all of these central variables now. 

The [garden client](https://pkg.go.dev/code.cloudfoundry.org/garden), which is use to fetch this information, does not provide any constants for these. As far as I can tell these are just properties created by [guardian](https://github.com/cloudfoundry/guardian/) when creating the containers, which doesn't seem to have them as constants either.